llvm/lib/SYCL/PrepareSYCLOpt.cpp

```diff
@@ -171,13 +173,50 @@ struct PrepareSYCLOpt : public ModulePass {
     }
   }
 
+  struct UnwrapperVisitor : public llvm::InstVisitor<UnwrapperVisitor> {
+    void visitCallInst(CallInst &I) {
+      auto *ParentF = I.getFunction();
+      auto *F = I.getCalledFunction();
+      if (!F->hasFnAttribute("fpga.propertywrapper"))
+        return;
+      // We have a property wrapper.
+      // First, unwrap all wrapper inside F
+      visit(*F);
+
+      // Now copy fpga attributes to parent
+      auto FnAttr = F->getAttributes().getFnAttributes();
+      for (auto &Attr : FnAttr) {
+        if (Attr.isStringAttribute()) {
+          StringRef AttrKind = Attr.getKindAsString();
+          if (AttrKind.startswith("fpga.") &&
+              AttrKind != "fpga.propertywrapper") {
+            ParentF->addFnAttr(Attr);
+          }
+        }
+      }
+      // And inline the wrapper inside the caller
```

Comment:
There should be a higher-level description somewhere,while I appreciate the local comments :-)